metaserver/contracts.go

```diff
@@ -183,3 +236,89 @@ func (server *MetaServer) deleteContractHandler() http.HandlerFunc {
 		w.WriteHeader(http.StatusOK)
 	})
 }
+
+func (server *MetaServer) getContractPaymentHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+
+		contract, err := server.db.FindContractByID(params["contractID"])
+		if err != nil {
+			writeErr(err.Error(), http.StatusNotFound, w)
+			return
+		}
+
+		// Make sure the person making the request is the renter who owns the files.
+		claims, err := util.GetTokenClaimsFromRequest(r)
+		if err != nil {
+			writeAndLogInternalError(err, w, server.logger)
+			return
+		}
+		if renterID, present := claims["renterID"]; !present || renterID.(string) != contract.RenterId {
+			writeErr("cannot retrieve other users' contracts", http.StatusUnauthorized, w)
+			return
+		}
+
+		// Retrieve the contract's payment information.
+		payment, err := server.db.FindPaymentByContract(contract.ID)
+		if err != nil {
+			writeAndLogInternalError(err, w, server.logger)
+			return
+		}
+
+		json.NewEncoder(w).Encode(payment)
+	})
+}
+
+func (server *MetaServer) putContractPaymentHandler() http.HandlerFunc {
```

Comment:
I'm curious why you didn't decide to go with a /cancel endpoint that takes no arguments. With this version, I see you have to perform some manual checks that the PUT contract matches the original. Perhaps this was more consistent with the rest of the API?

Reply:
That's exactly why I did that. Now that I'm looking at it again, I'm not liking it so much, but it does maintain consistency, so I think I'll just leave it for now.